tests/test_ostinato.py

```diff
@@ -137,3 +137,83 @@ def test_deterministic_ostinatoed_with_isconstant(seed, dask_cluster):
         npt.assert_almost_equal(ref_radius, comp_radius)
         npt.assert_almost_equal(ref_Ts_idx, comp_Ts_idx)
         npt.assert_almost_equal(ref_subseq_idx, comp_subseq_idx)
+
+
+@pytest.mark.parametrize(
+    "seed", np.random.choice(np.arange(10000), size=25, replace=False)
+)
+def test_input_not_overwritten(seed):
+    m = 50
+    np.random.seed(seed)
+    Ts = [np.random.rand(n) for n in [64, 128, 256]]
+
+    # without nan values
+    Ts_ref = [T.copy() for T in Ts]
+    Ts_comp = [T.copy() for T in Ts]
+    stumpy.ostinato(Ts_comp, m)
+    for i in range(len(Ts)):
+        npt.assert_almost_equal(Ts_ref[i], Ts_comp[i])
+
+    # with nan values
+    nan_size = [np.random.choice(np.arange(1, len(T) + 1)) for T in Ts]
+    for i in range(len(Ts)):
+        IDX = np.random.choice(np.arange(len(Ts[i])), size=nan_size[i], replace=False)
+        Ts[i][IDX] = np.nan
+
+    Ts_ref = [T.copy() for T in Ts]
+    Ts_comp = [T.copy() for T in Ts]
+    stumpy.ostinato(Ts_comp, m)
+    for i in range(len(Ts)):
+        T_ref = Ts_ref[i]
+        T_comp = Ts_comp[i]
+        npt.assert_almost_equal(T_ref[np.isfinite(T_ref)], T_comp[np.isfinite(T_comp)])
+
+
+@pytest.mark.parametrize(
+    "seed", np.random.choice(np.arange(10000), size=25, replace=False)
+)
+def test_extract_second_consensus(seed):
+    m = 50
+    np.random.seed(seed)
+    Ts = [np.random.rand(n) for n in [256, 512, 1024]]
+
+    # insert nan values
+    nan_size = [np.random.choice(np.arange(1, len(T) + 1)) for T in Ts]
+    for i in range(len(Ts)):
+        IDX = np.random.choice(np.arange(len(Ts[i])), size=nan_size[i], replace=False)
+        Ts[i][IDX] = np.nan
+
+    Ts_ref = [T.copy() for T in Ts]
+    Ts_comp = [T.copy() for T in Ts]
+
+    # obtain first consensus motif
+    central_radius, central_Ts_idx, central_subseq_idx = stumpy.ostinato(Ts_comp, m)
+
+    consensus_motif = Ts_comp[central_Ts_idx][
+        central_subseq_idx : central_subseq_idx + m
+    ].copy()
+    Ts_comp[central_Ts_idx][central_subseq_idx : central_subseq_idx + m] = np.nan
+    Ts_ref[central_Ts_idx][
+        central_subseq_idx : central_subseq_idx + m
+    ] = np.nan  # apply same changes to Ts_ref
+    for i in range(len(Ts)):
+        if i != central_Ts_idx:
+            D = stumpy.core.mass(consensus_motif, Ts_comp[i])
+            idx = np.argmin(D)
+            Ts_comp[i][idx : idx + m] = np.nan
+            Ts_ref[i][idx : idx + m] = np.nan  # apply same changes to Ts_ref
+
```

Comment:
So here, please assert that the original `Ts` is unchanged (i.e., it doesn't contain any `np.nan` or runs of `zero` that don't belong.

I'm not sure that the inserting of single (non-consecutive) `np.nan` above is necessary or what the goal is there?

Reply:
I was wondering if I should do the assertion right after passing `Ts_comp` to `ostinato`. I mean:

```
central_radius, central_Ts_idx, central_subseq_idx = stumpy.ostinato(Ts_comp, m)
for i in range(len(Ts)):
    ref = Ts[i]
    comp = Ts_comp[i]
    npt.assert_almost_equal(ref[np.isfinite(ref)], comp[np.isfinite(comp)])
```

Right? 
This way I can check if the original input `Ts_comp` remains untouched after calling `ostinato`. Particularly, to see if the fix works, I need to have at least one `np.nan` in the original data (so that it can be replaced with `0.0` in the local variable in the called function `ostinato` , and then we can check to see if that changes the original data or not ).  Btw, I checked this particular scenario in `test_input_not_overwritten`

---

To extract the second consensus, I DO CHANGE the  input `Ts_comp` here (by setting the consensus and its NN motifs to `np.nan`) before passing it to `ostinato` for the second time.

Reply:
Okay, I think this is fine then but please document the different/various intentions a little more clearly if possible

If it helps, it is important that the reason for each unit test is clear and, if it fails, it should be clear WHY it failed. We should avoid testing multiple different/unrelated things within the same unit test

Reply:
@seanlaw 
Right! I think things got mixed a bit there in my tests. Will add supportive comments and remove unnecessary tests.